src/test/java/com/haxwell/apps/quizki/services/QuestionServiceImplTest.java

```diff
@@ -0,0 +1,278 @@
+
+import static org.mockito.Mockito.*;
+import static org.mockito.BDDMockito.any;
+//Note this import added to avoid "any" naming collision in Mockito and Hamcrest Matchers see https://github.com/mockito/mockito/issues/1311
+
+import java.util.HashSet;
+import java.util.Optional;
+import java.util.Set;
+
+
+import static org.hamcrest.Matchers.*;
+import static org.junit.Assert.*;
+
+
+import org.junit.Before;
+import org.junit.Test;
+
+import com.haxwell.apps.quizki.dtos.CreateChoiceDTO;
+import com.haxwell.apps.quizki.dtos.CreateQuestionDTO;
+import com.haxwell.apps.quizki.dtos.CreatedQuestionDTO;
+import com.haxwell.apps.quizki.entities.Choice;
+import com.haxwell.apps.quizki.entities.Question;
+import com.haxwell.apps.quizki.entities.Reference;
+import com.haxwell.apps.quizki.entities.Topic;
+import com.haxwell.apps.quizki.entities.User;
+import com.haxwell.apps.quizki.entities.UserRole;
+import com.haxwell.apps.quizki.exceptions.CreateQuestionDTOException;
+import com.haxwell.apps.quizki.exceptions.GetQuestionException;
+import com.haxwell.apps.quizki.repositories.QuestionRepository;
+import com.haxwell.apps.quizki.repositories.ReferenceRepository;
+import com.haxwell.apps.quizki.repositories.TopicRepository;
+import com.haxwell.apps.quizki.repositories.UserRepository;
+
+public class QuestionServiceImplTest {
+	
+	QuestionServiceImpl qsImpl;
+	
+	CreateQuestionDTO inputDTO;
+	CreatedQuestionDTO outputDTO;
+	
+	Topic topic1;
+	Reference reference1;
+	Choice choice;
+	static User user;
+	static UserRole uRole;
+	static Question question;
+	
+	static String role1 = "QUIZKI_USER_ROLE_ADMIN";
+	static String role2 = "QUIZKI_USER_ROLE_USER";
+	
+	static Set<String> dtoTopics = new HashSet<String>();
+	static Set<String> dtoRefs = new HashSet<String>();
+	static Set<CreateChoiceDTO> dtoChoices = new HashSet<CreateChoiceDTO>();
+	
+	static Set<Topic> topics = new HashSet<Topic>();
+	static Set<Reference> references = new HashSet<Reference>();
+	static Set<Choice> choices = new HashSet<Choice>();
+	
+	static String topicStr1 = "topic1";
+	static String topicStr2 = "topic2";
+	
+	static String refStr1 = "reference1";
+	static String refStr2 = "reference2";
+	
+	static String text = "question text";
+	static String description = "question description";
+	
+	static long userId = 1;
+	static int type = 1;
+	static int difficulty = 1;
+	
+	static CreateChoiceDTO choiceDTO1T = new CreateChoiceDTO("choice1",true);
+	static CreateChoiceDTO choiceDTO2F = new CreateChoiceDTO("choice2", false);
+	
+	
+	@Mock
+	private QuestionRepository questionRepo;
+	
+	@Mock
+	private TopicRepository topicRepo;
+	
+	@Mock
+	private ReferenceRepository refRepo;
+		
+	@Mock
+	private UserRepository userRepo;
+	
+	@Before
+	public void setup() throws Exception {
+		MockitoAnnotations.initMocks(this);		
+		qsImpl = new QuestionServiceImpl(questionRepo, topicRepo, refRepo, userRepo);
+	}
+	
+	
+	private static CreateQuestionDTO getBaseInputDTO() {
+		
```

Comment:
A point on implementation... this method does more than return the BaseInputDTO.. it also initializes variables that are not defined in this method, specifically, the dto* objects. So, this method does two things; inits the dto objects, and returns an instance of CreateQuestionDTO. 

We should have two methods, one to init the values, and a second to return an instance of CreateQuestionDTO. You would pass in the dto* objects to the second method, and it would use those in the constructor.

Reply:
Actually, never mind! The method is only called from one place, so, I was jumping the gun. Optimizing before necessary. Never mind me!